utils/utils.go

```diff
@@ -14,101 +14,238 @@ limitations under the License.
 package utils
 
 import (
+	"archive/zip"
+	"io"
 	"io/ioutil"
+	"log"
 	"net/http"
+	"os"
+	"path/filepath"
 
 	pkgerrors "github.com/pkg/errors"
 	appsV1 "k8s.io/api/apps/v1"
 	coreV1 "k8s.io/api/core/v1"
 	"k8s.io/client-go/kubernetes/scheme"
 )
 
-// Download the raw CSAR file from a specific URL
-var Download = func(url string) ([]byte, error) {
+var CSAR FileOperator
+
+// FileOperator is an interface to Download, Extract and Delete files
+type FileOperator interface {
+	Download(string, string) error
+	Unzip(string) error
+	Delete(string) error
+}
+
+// CSARFile is a concrete type to Download, Extract and Delete CSAR files
+type CSARFile struct{}
+
+// Download a CSAR file
+func (c *CSARFile) Download(fileName string, url string) error {
+	file, err := os.Create(fileName)
+	if err != nil {
+		return pkgerrors.Wrap(err, "CSAR file create error")
+	}
+	defer file.Close()
+
 	resp, err := http.Get(url)
 	if err != nil {
-		return nil, pkgerrors.Wrap(err, "Get CSAR file error")
+		return pkgerrors.Wrap(err, "CSAR file download error")
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return pkgerrors.Wrap(err, "CSAR file download error")
 	}
 	defer resp.Body.Close()
 
-	rawBody, err := ioutil.ReadAll(resp.Body)
+	_, err = io.Copy(file, resp.Body)
+	if err != nil {
+		return pkgerrors.Wrap(err, "CSAR file encode error")
+	}
+
+	return nil
+}
+
+// Unzip a CSAR file
+func (c *CSARFile) Unzip(filePath string) error {
+	reader, err := zip.OpenReader(filePath)
 	if err != nil {
-		return nil, pkgerrors.Wrap(err, "Read CSAR file error")
+		return pkgerrors.Wrap(err, "CSAR file extracting error")
 	}
 
-	return rawBody, nil
+	for _, file := range reader.File {
+		path := filepath.Join(file.Name)
+
+		if file.FileInfo().IsDir() {
+			os.MkdirAll(path, file.Mode())
+			continue
+		}
+
+		fileReader, err := file.Open()
+		if err != nil {
+			return pkgerrors.Wrap(err, "CSAR file extracting error")
+		}
+		defer fileReader.Close()
+
+		targetFile, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, file.Mode())
+		if err != nil {
+			return pkgerrors.Wrap(err, "CSAR file extracting error")
+		}
+		defer targetFile.Close()
+
+		_, err = io.Copy(targetFile, fileReader)
+		if err != nil {
+			return pkgerrors.Wrap(err, "CSAR file extracting error")
+		}
+	}
+	return nil
 }
 
-// DownloadCSAR to download the CSAR files form URL and parse it
+// Delete a CSAR file
+func (c *CSARFile) Delete(path string) error {
+	err := os.RemoveAll(path)
+	if err != nil {
+		return pkgerrors.Wrap(err, "CSAR file delete error")
+	}
+	return nil
+}
+
+// CreateKubeObjectsFromCSAR to download the CSAR files from URL and extract it
 // to get deployment and service yamls
-var DownloadCSAR = func(url string) (*CSARData, error) {
-	rawFile, err := Download(url)
+var CreateKubeObjectsFromCSAR = func(csarID string, csarURL string) (*KubernetesData, error) {
```

Comment:
What about GetCSARFromURL instead of CreateKubeObjectsFromCSAR?

Reply:
Done.